flink-tests/src/test/java/org/apache/flink/test/misc/JobIDLoggingITCase.java

```diff
@@ -0,0 +1,227 @@
+import org.apache.flink.util.MdcUtils;
+import org.apache.flink.util.TestLogger;
+
+import org.apache.logging.log4j.core.LogEvent;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkState;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.slf4j.event.Level.DEBUG;
+
+/**
+ * Tests adding of {@link JobID} to logs (via {@link org.slf4j.MDC}) in the most important cases.
+ */
+public class JobIDLoggingITCase extends TestLogger {
+    private static final Logger logger = LoggerFactory.getLogger(JobIDLoggingITCase.class);
+
+    @RegisterExtension
+    public final LoggerAuditingExtension checkpointCoordinatorLogging =
+            new LoggerAuditingExtension(CheckpointCoordinator.class, DEBUG);
+
+    @RegisterExtension
+    public final LoggerAuditingExtension streamTaskLogging =
+            new LoggerAuditingExtension(StreamTask.class, DEBUG);
+
+    @RegisterExtension
+    public final LoggerAuditingExtension taskExecutorLogging =
+            new LoggerAuditingExtension(TaskExecutor.class, DEBUG);
+
+    @RegisterExtension
+    public final LoggerAuditingExtension taskLogging =
+            new LoggerAuditingExtension(Task.class, DEBUG);
+
+    @RegisterExtension
+    public final LoggerAuditingExtension executionGraphLogging =
+            new LoggerAuditingExtension(ExecutionGraph.class, DEBUG);
+
+    @RegisterExtension
+    public final LoggerAuditingExtension jobMasterLogging =
+            new LoggerAuditingExtension(JobMaster.class, DEBUG);
+
+    @RegisterExtension
+    public final LoggerAuditingExtension asyncCheckpointRunnableLogging =
+            // this class is private
+            new LoggerAuditingExtension(
+                    "org.apache.flink.streaming.runtime.tasks.AsyncCheckpointRunnable", DEBUG);
+
+    @RegisterExtension
+    public static MiniClusterExtension miniClusterResource =
+            new MiniClusterExtension(
+                    new MiniClusterResourceConfiguration.Builder()
+                            .setNumberTaskManagers(1)
+                            .setNumberSlotsPerTaskManager(1)
+                            .build());
+
+    @Test
+    public void testJobIDLogging(@InjectClusterClient ClusterClient<?> clusterClient)
+            throws Exception {
+        JobID jobID = null;
+        try {
+            jobID = runJob(clusterClient);
+        } finally {
+            if (jobID != null) {
+                clusterClient.cancel(jobID).get();
+            }
+        }
```

Comment:
```suggestion
        JobID jobID = runJob(clusterClient);
        clusterClient.cancel(jobID).get();
```
Maybe I'm missing something but the finally branch will never do anything in an exceptional case, because then the jobID couldn't have been assigned.